semver/parse_range.ts

```diff
@@ -234,34 +63,130 @@ function isWildcard(id: string): boolean {
  */
 export function parseRange(range: string): SemVerRange {
   // handle spaces around and between comparator and version
-  range = range.trim().replaceAll(/(?<=<|>|=) /g, "");
 
-  if (range === "") {
-    return { ranges: [[ALL]] };
-  }
+  if (range === "") return { ranges: [[ALL]] };
 
   // Split into groups of comparators, these are considered OR'd together.
   const ranges = range
-    .trim()
     .split(/\s*\|\|\s*/)
     .map((range) => {
-      range = hyphenReplace(range);
-      range = replaceCarets(range);
-      range = replaceTildes(range);
-      range = replaceXRanges(range);
-      range = replaceStars(range);
-
-      // At this point, the range is completely trimmed and
-      // ready to be split into comparators.
-      // These are considered AND's
-      if (range === "") {
-        return [ALL];
-      } else {
-        return range
-          .split(" ")
-          .filter((r) => r)
-          .map((r) => parseComparator(r));
-      }
+      // handle space between comparator and version
+      range = range.replace(/(?<=<|>|=) /, "");
+
+      const entries = hyphenReplace(range)
+        .flatMap((comp) => {
+          const groups = comp.match(OPERATOR_REGEXP)?.groups;
+          if (!groups) return comp.split(/\s+/);
+
+          const majorIsWildcard = isWildcard(groups.major);
+          const minorIsWildcard = isWildcard(groups.minor);
+          const patchIsWildcard = isWildcard(groups.patch);
+
+          const major = +groups.major;
+          const minor = +groups.minor;
+          const patch = +groups.patch;
+
+          switch (groups.operator) {
```

Comment:
This `switch` statement is quite meaty and difficult to follow. I liked that the previous approach factored out this logic into separate functions. Maybe we can factor out the logic of each case as before but still keep the other cleanups in this PR. WDYT?

Reply:
Yeah, I agree. I thought about that too and figured separate functions would span lots of duplicate code, especially the `majorIsWildcard` checks and number type castings, nut I think it looks good enough and updated the PR.
**Note** I replaced all string return values with a `parseComparator()` wrapper, so the return type of the functions, so that these functions don't return an array of strings which have to be parsed later. We can replace these wrappers with the actual Operator `{ operator: ..., semver: ...}` value in a later PR, so no parsing step is needed.